rpcs3/Emu/SysCalls/Modules/cellPngDec.cpp

```diff
@@ -3,572 +3,859 @@
 	{
-		auto file = idm::get<lv2_file_t>(stream->fd);
-		file->file->Seek(0);
-		file->file->Read(png.get(), fileSize);
-		break;
+		// Check if the packing is really 1 byte per 1 pixel
+		if (stream->packing != CELL_PNGDEC_1BYTE_PER_1PIXEL)
+		{
+			throw EXCEPTION("Unexpected packing value! (%d)", stream->packing);
+		}
+
+		// Scale 16 bit depth down to 8 bit depth. PS3 uses png_set_strip_16, since png_set_scale_16 wasn't available back then.
+		png_set_strip_16(stream->png_ptr);
+
+		// We need to tell libpng to update the info structure, since we modified the info
+		png_read_update_info(stream->png_ptr, stream->info_ptr);
+
+		// Recalculate the outputWidthByte value, and recalculate the image size
+		stream->out_param.outputWidthByte = png_get_rowbytes(stream->png_ptr, stream->info_ptr);
+
+		// What do we do if bytes_per_line is still samller? Something probably went wrong, but this should never happen anyway.
+		if (bytes_per_line < stream->out_param.outputWidthByte)
+		{
+			throw EXCEPTION("bytesPerLine is still smaller than outputWidthByte: %d", stream->out_param.outputWidthByte);
+		}
+	}
+	// Check if the outputWidthByte is smaller than the intended output length of a line. For example an image might be in RGB, but we need to output 4 components, so we need to perform alpha padding.
+	else if (stream->out_param.outputWidthByte < (stream->out_param.outputWidth * stream->out_param.outputComponents))
+	{
+		// Not sure what to do, when a fixed alpha value isn't specified.
+		if (!stream->fixed_alpha)
+		{
+			throw EXCEPTION("Fixed alpha not specified for padding.");
+		}
+
+		// We need to fill alpha (before or after, depending on the output colour format) using the fixed alpha value passed by the game.
+		png_set_add_alpha(stream->png_ptr, stream->fixed_alpha_colour, stream->out_param.outputColorSpace == CELL_PNGDEC_RGBA ? PNG_FILLER_AFTER : PNG_FILLER_BEFORE);
+		
+		// We need to tell libpng to update the info structure, since we modified the info
+		png_read_update_info(stream->png_ptr, stream->info_ptr);
+
+		// Calculate the actual needed output image size
+		stream->out_param.outputWidthByte = stream->out_param.outputWidth * stream->out_param.outputComponents;
 	}
+	else if (stream->out_param.outputColorSpace == CELL_PNGDEC_ARGB)
+	{
+		// Swap the alpha channel for the ARGB output format, if the padding isn't needed
+		png_set_swap_alpha(stream->png_ptr);
 	}
 
-	//Decode PNG file. (TODO: Is there any faster alternative? Can we do it without external libraries?)
-	int width, height, actual_components;
-	auto image = std::unique_ptr<unsigned char, decltype(&::free)>
-		(
-		stbi_load_from_memory(png.get(), (s32)fileSize, &width, &height, &actual_components, 4),
-		&::free
-		);
-	if (!image)
+	// Calculate the image size
+	u32 image_size = stream->out_param.outputWidthByte * stream->out_param.outputHeight;
+
+	// Buffer for storing the image
+	std::unique_ptr<u8[]> png(new u8[image_size]);
```

Comment:
I don't think the function will be called that often. Decoding a png is costly.
If you really don't want to pay initialisation cost then use gsl::span<u8> so that some bound checking can occur here :
 "row_pointers.get()[y] = &png.get()[y * stream->out_param.outputWidthByte];"